File: output/cpp/tests/20_transition_pop.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <any>
#include <memory>


#include <iostream>
#include <string>
#include <cassert>
#include <vector>
#include <algorithm>

class TransitionPopTestFrameEvent {
public:
    std::string _message;
    std::unordered_map<std::string, std::any> _parameters;

    TransitionPopTestFrameEvent(const std::string& message, std::unordered_map<std::string, std::any> params = {})
        : _message(message), _parameters(std::move(params)) {}
};

class TransitionPopTestFrameContext {
public:
    TransitionPopTestFrameEvent _event;
    std::any _return;
    std::unordered_map<std::string, std::any> _data;

    TransitionPopTestFrameContext(TransitionPopTestFrameEvent event, std::any default_return = {})
        : _event(std::move(event)), _return(std::move(default_return)) {}
};

class TransitionPopTestCompartment {
public:
    std::string state;
    std::unordered_map<std::string, std::any> state_args;
    std::unordered_map<std::string, std::any> state_vars;
    std::unordered_map<std::string, std::any> enter_args;
    std::unordered_map<std::string, std::any> exit_args;
    std::unique_ptr<TransitionPopTestFrameEvent> forward_event;
    std::unique_ptr<TransitionPopTestCompartment> parent_compartment;

    explicit TransitionPopTestCompartment(const std::string& state) : state(state) {}

    std::unique_ptr<TransitionPopTestCompartment> clone() const {
        auto c = std::make_unique<TransitionPopTestCompartment>(state);
        c->state_args = state_args;
        c->state_vars = state_vars;
        c->enter_args = enter_args;
        c->exit_args = exit_args;
        return c;
    }
};

class TransitionPopTest {
private:
    std::vector<std::unique_ptr<TransitionPopTestCompartment>> _state_stack;
    std::unique_ptr<TransitionPopTestCompartment> __compartment;
    std::unique_ptr<TransitionPopTestCompartment> __next_compartment;
    std::vector<TransitionPopTestFrameContext> _context_stack;

    void __kernel(TransitionPopTestFrameEvent& __e) {
        __router(__e);
        while (__next_compartment) {
            auto next_compartment = std::move(__next_compartment);
            TransitionPopTestFrameEvent exit_event("<$");
            __router(exit_event);
            __compartment = std::move(next_compartment);
            if (!__compartment->forward_event) {
                TransitionPopTestFrameEvent enter_event("$>");
                __router(enter_event);
            } else {
                auto forward_event = std::move(__compartment->forward_event);
                if (forward_event->_message == "$>") {
                    __router(*forward_event);
                } else {
                    TransitionPopTestFrameEvent enter_event("$>");
                    __router(enter_event);
                    __router(*forward_event);
                }
            }
        }
    }

    void __router(TransitionPopTestFrameEvent& __e) {
        const auto& state_name = __compartment->state;
        if (state_name == "Idle") {
            _state_Idle(__e);
        } else if (state_name == "Working") {
            _state_Working(__e);
        }
    }

    void __transition(std::unique_ptr<TransitionPopTestCompartment> next) {
        __next_compartment = std::move(next);
    }

    void _state_Idle(TransitionPopTestFrameEvent& __e) {
        if (__e._message == "get_log") {
            _context_stack.back()._return = std::any(event_log);
            return;;
        } else if (__e._message == "get_state") {
            _context_stack.back()._return = std::any(std::string("Idle"));
            return;;
        } else if (__e._message == "process") {
            event_log.push_back("idle:process");
        } else if (__e._message == "start") {
            event_log.push_back("idle:start:push");
            _state_stack.push_back(__compartment->clone());
            auto __new_compartment = std::make_unique<TransitionPopTestCompartment>("Working");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        }
    }

    void _state_Working(TransitionPopTestFrameEvent& __e) {
        if (__e._message == "get_log") {
            _context_stack.back()._return = std::any(event_log);
            return;;
        } else if (__e._message == "get_state") {
            _context_stack.back()._return = std::any(std::string("Working"));
            return;;
        } else if (__e._message == "process") {
            event_log.push_back("working:process:before_pop");
            auto __popped = std::move(_state_stack.back()); _state_stack.pop_back();
            __transition(std::move(__popped));
            return;
            // This should NOT execute because pop transitions away
            event_log.push_back("working:process:after_pop");
        }
    }

public:
    std::vector<std::string> event_log = {};

    TransitionPopTest() {
        __compartment = std::make_unique<TransitionPopTestCompartment>("Idle");
        TransitionPopTestFrameEvent __frame_event("$>");
        TransitionPopTestFrameContext __ctx(std::move(__frame_event));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void start() {
        TransitionPopTestFrameEvent __e("start");
        TransitionPopTestFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void process() {
        TransitionPopTestFrameEvent __e("process");
        TransitionPopTestFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    std::string get_state() {
        TransitionPopTestFrameEvent __e("get_state");
        TransitionPopTestFrameContext __ctx(std::move(__e), std::any(std::string()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<std::string>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    std::vector<std::string> get_log() {
        TransitionPopTestFrameEvent __e("get_log");
        TransitionPopTestFrameContext __ctx(std::move(__e), std::any(std::vector<std::string>()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<std::vector<std::string>>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }
};

int main() {
    printf("=== Test 20: Transition Pop (C++) ===\n");
    TransitionPopTest s;

    if (s.get_state() != "Idle") {
        printf("FAIL: Expected 'Idle', got '%s'\n", s.get_state().c_str());
        assert(false);
    }
    printf("Initial state: %s\n", s.get_state().c_str());

    s.start();
    if (s.get_state() != "Working") {
        printf("FAIL: Expected 'Working', got '%s'\n", s.get_state().c_str());
        assert(false);
    }
    printf("After start(): %s\n", s.get_state().c_str());

    s.process();
    if (s.get_state() != "Idle") {
        printf("FAIL: Expected 'Idle' after pop, got '%s'\n", s.get_state().c_str());
        assert(false);
    }
    printf("After process() with pop: %s\n", s.get_state().c_str());

    auto log = s.get_log();
    if (std::find(log.begin(), log.end(), "idle:start:push") == log.end()) {
        printf("FAIL: Expected 'idle:start:push' in log\n");
        assert(false);
    }
    if (std::find(log.begin(), log.end(), "working:process:before_pop") == log.end()) {
        printf("FAIL: Expected 'working:process:before_pop' in log\n");
        assert(false);
    }
    if (std::find(log.begin(), log.end(), "working:process:after_pop") != log.end()) {
        printf("FAIL: Should NOT have 'working:process:after_pop' in log\n");
        assert(false);
    }

    printf("PASS: Transition pop works correctly\n");
    return 0;
}

File: output/csharp/tests/20_transition_pop.cs
using System;
using System.Collections.Generic;


using System;
using System.Collections.Generic;

class TransitionPopTestFrameEvent {
    public string _message;
    public Dictionary<string, object> _parameters;

    public TransitionPopTestFrameEvent(string message) {
        this._message = message;
        this._parameters = null;
    }

    public TransitionPopTestFrameEvent(string message, Dictionary<string, object> parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}

class TransitionPopTestFrameContext {
    public TransitionPopTestFrameEvent _event;
    public object _return;
    public Dictionary<string, object> _data;

    public TransitionPopTestFrameContext(TransitionPopTestFrameEvent ev, object defaultReturn) {
        this._event = ev;
        this._return = defaultReturn;
        this._data = new Dictionary<string, object>();
    }
}

class TransitionPopTestCompartment {
    public string state;
    public Dictionary<string, object> state_args;
    public Dictionary<string, object> state_vars;
    public Dictionary<string, object> enter_args;
    public Dictionary<string, object> exit_args;
    public TransitionPopTestFrameEvent forward_event;
    public TransitionPopTestCompartment parent_compartment;

    public TransitionPopTestCompartment(string state) {
        this.state = state;
        this.state_args = new Dictionary<string, object>();
        this.state_vars = new Dictionary<string, object>();
        this.enter_args = new Dictionary<string, object>();
        this.exit_args = new Dictionary<string, object>();
        this.forward_event = null;
        this.parent_compartment = null;
    }

    public TransitionPopTestCompartment Copy() {
        TransitionPopTestCompartment c = new TransitionPopTestCompartment(this.state);
        c.state_args = new Dictionary<string, object>(this.state_args);
        c.state_vars = new Dictionary<string, object>(this.state_vars);
        c.enter_args = new Dictionary<string, object>(this.enter_args);
        c.exit_args = new Dictionary<string, object>(this.exit_args);
        c.forward_event = this.forward_event;
        c.parent_compartment = this.parent_compartment;
        return c;
    }
}

class TransitionPopTest {
    private List<TransitionPopTestCompartment> _state_stack;
    private TransitionPopTestCompartment __compartment;
    private TransitionPopTestCompartment __next_compartment;
    private List<TransitionPopTestFrameContext> _context_stack;
    public List<object> log = new List<object>();

    public TransitionPopTest() {
        _state_stack = new List<TransitionPopTestCompartment>();
        _context_stack = new List<TransitionPopTestFrameContext>();
        __compartment = new TransitionPopTestCompartment("Idle");
        __next_compartment = null;
        TransitionPopTestFrameEvent __frame_event = new TransitionPopTestFrameEvent("$>");
        TransitionPopTestFrameContext __ctx = new TransitionPopTestFrameContext(__frame_event, null);
        _context_stack.Add(__ctx);
        __kernel(_context_stack[_context_stack.Count - 1]._event);
        _context_stack.RemoveAt(_context_stack.Count - 1);
    }

    private void __kernel(TransitionPopTestFrameEvent __e) {
        __router(__e);
        while (__next_compartment != null) {
            var next_compartment = __next_compartment;
            __next_compartment = null;
            TransitionPopTestFrameEvent exit_event = new TransitionPopTestFrameEvent("<$");
            __router(exit_event);
            __compartment = next_compartment;
            if (__compartment.forward_event == null) {
                TransitionPopTestFrameEvent enter_event = new TransitionPopTestFrameEvent("$>");
                __router(enter_event);
            } else {
                var forward_event = __compartment.forward_event;
                __compartment.forward_event = null;
                if (forward_event._message == "$>") {
                    __router(forward_event);
                } else {
                    TransitionPopTestFrameEvent enter_event = new TransitionPopTestFrameEvent("$>");
                    __router(enter_event);
                    __router(forward_event);
                }
            }
        }
    }

    private void __router(TransitionPopTestFrameEvent __e) {
        string state_name = __compartment.state;
        if (state_name == "Idle") {
            _state_Idle(__e);
        } else if (state_name == "Working") {
            _state_Working(__e);
        }
    }

    private void __transition(TransitionPopTestCompartment next) {
        __next_compartment = next;
    }

    public void start() {
        TransitionPopTestFrameEvent __e = new TransitionPopTestFrameEvent("start");
        TransitionPopTestFrameContext __ctx = new TransitionPopTestFrameContext(__e, null);
        _context_stack.Add(__ctx);
        __kernel(_context_stack[_context_stack.Count - 1]._event);
        _context_stack.RemoveAt(_context_stack.Count - 1);
    }

    public void process() {
        TransitionPopTestFrameEvent __e = new TransitionPopTestFrameEvent("process");
        TransitionPopTestFrameContext __ctx = new TransitionPopTestFrameContext(__e, null);
        _context_stack.Add(__ctx);
        __kernel(_context_stack[_context_stack.Count - 1]._event);
        _context_stack.RemoveAt(_context_stack.Count - 1);
    }

    public string get_state() {
        TransitionPopTestFrameEvent __e = new TransitionPopTestFrameEvent("get_state");
        TransitionPopTestFrameContext __ctx = new TransitionPopTestFrameContext(__e, null);
        _context_stack.Add(__ctx);
        __kernel(_context_stack[_context_stack.Count - 1]._event);
        var __result = (string) _context_stack[_context_stack.Count - 1]._return;
        _context_stack.RemoveAt(_context_stack.Count - 1);
        return __result;
    }

    public List<object> get_log() {
        TransitionPopTestFrameEvent __e = new TransitionPopTestFrameEvent("get_log");
        TransitionPopTestFrameContext __ctx = new TransitionPopTestFrameContext(__e, null);
        _context_stack.Add(__ctx);
        __kernel(_context_stack[_context_stack.Count - 1]._event);
        var __result = (List<object>) _context_stack[_context_stack.Count - 1]._return;
        _context_stack.RemoveAt(_context_stack.Count - 1);
        return __result;
    }

    private void _state_Idle(TransitionPopTestFrameEvent __e) {
        if (__e._message == "get_log") {
            _context_stack[_context_stack.Count - 1]._return = this.log;
            return;
        } else if (__e._message == "get_state") {
            _context_stack[_context_stack.Count - 1]._return = "Idle";
            return;
        } else if (__e._message == "process") {
            this.log.Add("idle:process");
        } else if (__e._message == "start") {
            this.log.Add("idle:start:push");
            _state_stack.Add(__compartment.Copy());
            { var __new_compartment = new TransitionPopTestCompartment("Working");
            __new_compartment.parent_compartment = __compartment.Copy();
            __transition(__new_compartment); }
            return;
        }
    }

    private void _state_Working(TransitionPopTestFrameEvent __e) {
        if (__e._message == "get_log") {
            _context_stack[_context_stack.Count - 1]._return = this.log;
            return;
        } else if (__e._message == "get_state") {
            _context_stack[_context_stack.Count - 1]._return = "Working";
            return;
        } else if (__e._message == "process") {
            this.log.Add("working:process:before_pop");
            var __popped = _state_stack[_state_stack.Count - 1]; _state_stack.RemoveAt(_state_stack.Count - 1);
            __transition(__popped);
            return;
        }
    }
}

class Program {
    static void Main(string[] args) {
        Console.WriteLine("=== Test 20: Transition Pop (C#) ===");
        var s = new TransitionPopTest();

        // Initial state should be Idle
        if ((string)s.get_state() != "Idle") {
            Console.WriteLine("FAIL: Expected 'Idle', got '" + s.get_state() + "'");
            Environment.Exit(1);
        }
        Console.WriteLine("Initial state: " + s.get_state());

        // start() pushes Idle, transitions to Working
        s.start();
        if ((string)s.get_state() != "Working") {
            Console.WriteLine("FAIL: Expected 'Working', got '" + s.get_state() + "'");
            Environment.Exit(1);
        }
        Console.WriteLine("After start(): " + s.get_state());

        // process() in Working does pop transition back to Idle
        s.process();
        if ((string)s.get_state() != "Idle") {
            Console.WriteLine("FAIL: Expected 'Idle' after pop, got '" + s.get_state() + "'");
            Environment.Exit(1);
        }
        Console.WriteLine("After process() with pop: " + s.get_state());

        var log = (List<object>)s.get_log();
        Console.WriteLine("Log: [" + string.Join(", ", log) + "]");

        // Verify log contents
        if (!log.Contains("idle:start:push")) {
            Console.WriteLine("FAIL: Expected 'idle:start:push' in log");
            Environment.Exit(1);
        }
        if (!log.Contains("working:process:before_pop")) {
            Console.WriteLine("FAIL: Expected 'working:process:before_pop' in log");
            Environment.Exit(1);
        }
        if (log.Contains("working:process:after_pop")) {
            Console.WriteLine("FAIL: Should NOT have 'working:process:after_pop' in log");
            Environment.Exit(1);
        }

        Console.WriteLine("PASS: Transition pop works correctly");
    }
}
